paddle/majel/dim.h

```diff
@@ -0,0 +1,451 @@
+#pragma once
+
+#include <iostream>
+#include <sstream>
+#include <stdexcept>
+#include <type_traits>
+
+#include "majel/hostdevice.h"
+#include "majel/util.h"
```

Comment:
How about moving this header file `majel/util.h` into `majel/detail/assert.h`?  Please see my comments in this file for more details.